src/proof.rs

```diff
@@ -0,0 +1,170 @@
+
+use crypto::digest::Digest;
+
+use hashable::{ Hashable };
+use merkledigest::{ MerkleDigest };
+use merkletree::{ MerkleTree };
+use tree::{ Tree };
+
+/// An inclusion proof represent the fact that `value` is a member of a `MerkleTree`
+/// with root hash `root_hash`, and hash function `digest`.  
+/// A proof is a linked-list of `ProofBlock`s.  
+/// TODO: Represent a proof as a vector of ProofBlock instead of a linked-list?
+pub struct Proof<D, T> {
+    pub digest: D,
+    pub root_hash: Vec<u8>,
+    pub block: ProofBlock,
+    pub value: T
+}
+
+impl <D, T> Proof<D, T> where D: Digest + Clone, T: Hashable + Clone {
+
+    // TODO: Refactor this mess (@romac)
+    pub fn validate_against(&self, tree: &mut MerkleTree<D, T>) -> bool {
```

Comment:
This doesn't make much sense to me.

The problem we're trying to address is that a given value was in the tree. But we don't have the full tree, otherwise we could just verify it ourselves and it wouldn't be of any value to verify a proof given by a third party.

Instead, all I know is the **root** of a given Merkle tree but not the full tree. Verifying a proof means verifying that:
- the ProofBlock I'm given is valid (respects the structure of the Merkle tree where parent = H(Left + Right)
- the leaf is the hash and value of the data is the one I'm being given an inclusion proof of
- the root of the Merkle tree is the same than the one I know of

Makes sense?